client/network/src/protocol/generic_proto/handler.rs

```diff
@@ -765,28 +765,13 @@ impl ProtocolsHandler for NotifsHandler {
 			}
 
 			// Poll outbound substream.
-			match &mut self.protocols[protocol_index].state {
-				State::Open { out_substream: out_substream @ Some(_), .. } => {
-					match Sink::poll_flush(Pin::new(out_substream.as_mut().unwrap()), cx) {
-						Poll::Pending | Poll::Ready(Ok(())) => {},
-						Poll::Ready(Err(_)) => {
-							*out_substream = None;
-							let event = NotifsHandlerOut::CloseDesired { protocol_index };
-							return Poll::Ready(ProtocolsHandlerEvent::Custom(event));
-						}
-					};
-				}
-
-				State::Closed { .. } |
-				State::Opening { .. } |
-				State::Open { out_substream: None, .. } |
-				State::OpenDesiredByRemote { .. } => {}
-			}
-
-			if let State::Open { notifications_sink_rx, out_substream: Some(out_substream), .. }
+			if let State::Open { notifications_sink_rx, out_substream: out_substream @ Some(_), .. }
 				= &mut self.protocols[protocol_index].state
 			{
+				let mut flush_needed = false;
```

Comment:
Since this is local, if a flush is needed but `poll_flush` returns `Poll::Pending`, it looks like the handler may not call `poll_flush()` again when it is next polled? I would think that not wanting to track such state is why the current code always calls `poll_flush()` for each open substream in the removed block above. Please correct me if I'm mistaken.

Reply:
Good point!

The problem that we have encountered is that flush wasn't called at all, we assume that this was probably due to other substreams being more busy or whatever.

As I'm thinking more about this, the current implementation isn't really robust. It could happen that a stream with a lower `procotol_index` could starve other streams. There is already this `events_queue`. We should insert every event in there and to ensure that we poll all incoming and outgoing streams. At the end of this loop we should check if there is something in the `events_queue` and return that.

Reply:
Good catch @romanb ! Thanks!

Reply:
> As I'm thinking more about this, the current implementation isn't really robust. It could happen that a stream with a lower `procotol_index` could starve other streams. There is already this `events_queue`. We should insert every event in there and to ensure that we poll all incoming and outgoing streams. At the end of this loop we should check if there is something in the `events_queue` and return that.

Problem: We lose back pressure. `events_queue` could grow indefinitely in that case.

Reply:
Why do we loose the back pressure? The event queue was checked before we polled for new events?

Reply:
Yeah, I wanted to avoid checking before _and_ after and if we only check afterwards, we got that issue. If we only checked first, we would return `Pending` although we are actually `Ready`. Either way @tomaka seems skeptic that we solve anything here.